### PR TITLE
[FLINK-14593][client] Port ClusterClient to asynchronous interface version

### PR DESCRIPTION
**Pull request by TisonKun**
## What is the purpose of the change

Port ClusterClient to asynchronous interface version

## Brief change log

1. Port ClusterClient to interface
2. Port `cancel`/`cancelWithSavepoint`/`stopWithSavepoint`/`getAccumulators` to asynchronous version
3. Introduce a settable `TestingClusterClient` and remove `FakeClusterClient`
4. Adjust related code.

## Verifying this change

This change is a refactor and covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by TisonKun** [comment]
Hi @kl0u @aljoscha could you take a look at this?

File: flink-clients/src/main/java/org/apache/flink/client/program/MiniClusterClient.java [hunk 1]
```diff
@@ -120,7 +135,25 @@ public String stopWithSavepoint(JobID jobId, boolean advanceToEndOfEventTime, @N
...
+		} catch (InterruptedException | ExecutionException e) {
+			ExceptionUtils.checkInterrupted(e);
+
+			LOG.warn("Could not retrieve the web interface URL for the cluster.", e);
+			return "Unknown address.";
+		}
+	}
+
+	@Override
+	public void shutDownCluster() {
```
Comment (kl0u) [inline flink-clients/src/main/java/org/apache/flink/client/program/MiniClusterClient.java]: This seems correct but it is a change of behaviour as previously we were simply throwing an `UnsupportedOperationException`. Given this, it deserves a separate JIRA issue and separate commit.

File: flink-clients/src/test/java/org/apache/flink/client/cli/util/MockedCliFrontend.java [hunk 2]
```diff
@@ -30,8 +30,7 @@
  * <p>The mocking behavior can be defined in the constructor of the sub-class.
  */
 public class MockedCliFrontend extends CliFrontend {
-
```
Comment (kl0u) [inline flink-clients/src/test/java/org/apache/flink/client/cli/util/MockedCliFrontend.java]: Please re-add the empty line.

File: flink-clients/src/test/java/org/apache/flink/client/cli/CliFrontendStopWithSavepointTest.java [hunk 3]
```diff
@@ -67,78 +62,119 @@ public void testStopWithOnlyJobId() throws Exception {
...
 	@Test
 	public void testStopOnlyWithMaxWM() throws Exception {
 		JobID jid = new JobID();
 
 		String[] parameters = { "-d", jid.toString() };
-		final ClusterClient<String> clusterClient = createClusterClient(null);
+		OneShotLatch stopWithSavepointLatch = new OneShotLatch();
+		TestingClusterClient<String> clusterClient = new TestingClusterClient<>();
+		clusterClient.setStopWithSavepointFunction((jobID, advanceToEndOfEventTime, savepointDirectory) -> {
+				assertThat(jobID, is(jid));
```
Comment (kl0u) [inline flink-clients/src/test/java/org/apache/flink/client/cli/CliFrontendStopWithSavepointTest.java]: Remove a tab from all these lines for uniformity.

File: flink-tests/src/test/java/org/apache/flink/test/accumulators/AccumulatorLiveITCase.java [hunk 4]
```diff
@@ -160,7 +160,7 @@ public void go() throws Exception {
 			FutureUtils.retrySuccessfulWithDelay(
 				() -> {
 					try {
-						return CompletableFuture.completedFuture(client.getAccumulators(jobGraph.getJobID()));
+						return CompletableFuture.completedFuture(client.getAccumulators(jobGraph.getJobID())).get();
```
Comment (kl0u) [inline flink-tests/src/test/java/org/apache/flink/test/accumulators/AccumulatorLiveITCase.java]: I think the `get()` should be in the parenthesis: 
```
CompletableFuture.completedFuture(client.getAccumulators(jobGraph.getJobID())).get()
```

Right?

File: flink-clients/src/main/java/org/apache/flink/client/program/MiniClusterClient.java (hunk 1 above)
Comment (TisonKun) [reply flink-clients/src/main/java/org/apache/flink/client/program/MiniClusterClient.java]: OK. I will add a default "implementation" throw an `UnsupportedOperationException` in this pass.

File: flink-clients/src/test/java/org/apache/flink/client/cli/CliFrontendStopWithSavepointTest.java (hunk 3 above)
Comment (TisonKun) [reply flink-clients/src/test/java/org/apache/flink/client/cli/CliFrontendStopWithSavepointTest.java]: Thanks!

File: flink-tests/src/test/java/org/apache/flink/test/accumulators/AccumulatorLiveITCase.java (hunk 4 above)
Comment (TisonKun) [reply flink-tests/src/test/java/org/apache/flink/test/accumulators/AccumulatorLiveITCase.java]: Thanks for spotting this. Correctly we should `get` on the inner future so that we get a completed future.

File: flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java [hunk 5]
```diff
@@ -704,17 +704,14 @@ private String triggerSavepoint(ClusterClient<?> clusterClient, JobID jobId, Str
 		final String savepointPath;
 
 		try {
-			savepointPath = savepointPathFuture.get();
-		}
-		catch (Exception e) {
+			savepointPath = savepointPathFuture.get(clientTimeout.toMillis(), TimeUnit.MILLISECONDS);
```
[the remark this answers is not in the document]
Comment (TisonKun) [reply flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java]: make sense.

**Comment by TisonKun** [comment]
Thanks for your review @kl0u ! I've updated the pull request.

**Comment by TisonKun** [comment]
@flinkbot run travis

**Comment by TisonKun** [comment]
Thanks for your review @kl0u !
